cmd/tlssocks/tlssocks.go

```diff
@@ -134,8 +128,7 @@ func (s Credentials) Valid(user, password string) bool {
 	if !ok {
 		return false
 	}
-	errHash := bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(password))
-	return errHash == nil
+	return nil == bcrypt.CompareHashAndPassword([]byte(hashedPassword), []byte(password))
```

Comment:
Discussion: What's the speed of compare hash and password? Like does it add a 100ms overhead or something if the hash algorithm is high?

Reply:
It's important to know due to CPU intense operations if we have a lot of requests coming in that it's not a bottleneck.

Reply:
You'll expose yourself to timing attacks if you cache this operation.

Reply:
@urandom actually, quite the opposite! The timing attack requires discrepancies in the calculation part.
So, the more the calculation, the more the time discrepancy, the more the chance of a timing attack, if for instance we store the passwords there.

Currently we're vulnerable to a timing attack "at least for a username" since it answers faster if the username is invalid than the password.

ALL this said, we could do a Certificate Authentication instead, or just put up a firewall for all intents and purposes, thus potentially saving the requirement for a large machine, depending on the traffic.

Would be immensely useful if we had metrics on the k8s side for the response times!

Reply:
refactored with caching. please review @smartinov @urandom